wagtail/wagtailadmin/views/pages.py

```diff
@@ -542,133 +544,96 @@ def view_draft(request, page_id):
     return page.serve_preview(page.dummy_request(request), page.default_preview_mode)
 
 
-def preview_on_edit(request, page_id):
-    # Receive the form submission that would typically be posted to the 'edit' view. If submission is valid,
-    # return the rendered page; if not, re-render the edit form
-    page = get_object_or_404(Page, id=page_id).get_latest_revision_as_page()
-    content_type = page.content_type
-    page_class = content_type.model_class()
-    parent_page = page.get_parent().specific
-    edit_handler_class = page_class.get_edit_handler()
-    form_class = edit_handler_class.get_form_class(page_class)
-
-    form = form_class(request.POST, request.FILES, instance=page, parent_page=parent_page)
-
-    if form.is_valid():
-        form.save(commit=False)
-        page.full_clean()
-
-        preview_mode = request.GET.get('mode', page.default_preview_mode)
-        response = page.serve_preview(page.dummy_request(request), preview_mode)
-        response['X-Wagtail-Preview'] = 'ok'
-        return response
-
-    else:
-        edit_handler = edit_handler_class(instance=page, form=form)
-
-        response = render(request, 'wagtailadmin/pages/edit.html', {
-            'page': page,
-            'edit_handler': edit_handler,
-            'preview_modes': page.preview_modes,
-            'form': form,
-            'has_unsaved_changes': True,
-        })
-        response['X-Wagtail-Preview'] = 'error'
-        return response
-
-
-def preview_on_create(request, content_type_app_name, content_type_model_name, parent_page_id):
-    # Receive the form submission that would typically be posted to the 'create' view. If submission is valid,
-    # return the rendered page; if not, re-render the edit form
-    try:
-        content_type = ContentType.objects.get_by_natural_key(content_type_app_name, content_type_model_name)
-    except ContentType.DoesNotExist:
-        raise Http404
-
-    page_class = content_type.model_class()
-    page = page_class()
-    edit_handler_class = page_class.get_edit_handler()
-    form_class = edit_handler_class.get_form_class(page_class)
-    parent_page = get_object_or_404(Page, id=parent_page_id).specific
-
-    form = form_class(request.POST, request.FILES, instance=page, parent_page=parent_page)
+class PreviewOnEdit(View):
+    http_method_names = ('post', 'get')
+    preview_expiration_timeout = 60 * 60 * 24  # seconds
```

Comment:
It’s not really important, in fact. Every time the preview for a page is updated, the previous preview data is removed. That timeout is used to delete old previews for other pages. If you create a preview for page 1, wait 10 minutes, work on page 2 and make a preview of it, maybe you kept preview for page 1 in a tab. So it’s better to keep it for at least a few hours in my opinion. When you create a preview for page 2 more than 24 hours after a page 1 preview, then data from this preview 1 is deleted.